ocaml/xapi/xapi_session.ml

```diff
@@ -27,6 +27,228 @@ open Client
+  }
+
+  let string_of_stats
+      {
+        total_num_failed_attempts
+      ; top_3_worst_clients
+      ; unknown_client_failed_attempts
+      } =
+    let string_of_client_failed_attempts x =
+      Printf.sprintf "- %i failed login attempts! Last failed attempt: %s, %s"
+        x.num_failed_attempts
+        (Date.to_string x.last_failed_attempt)
+        (string_of_client x.client)
+    in
+    match top_3_worst_clients with
+    | [] ->
+        Printf.sprintf
+          "We found a total of %i failed attempts to login since last checking \
+           - these were all of unknown origin!"
+          unknown_client_failed_attempts
+    | _ :: _ ->
+        let worst =
+          List.map string_of_client_failed_attempts top_3_worst_clients
+          |> String.concat "\n"
+        in
+        if unknown_client_failed_attempts > 0 then
+          Printf.sprintf
+            "We found a total of %i failed attempts to login since last \
+             checking, including %i of unknown origin! The worst known \
+             offenders were:\n\
+             %s"
+            total_num_failed_attempts unknown_client_failed_attempts worst
+        else
+          Printf.sprintf
+            "We found a total of %i failed attempts to login since last \
+             checking! The worst offenders were:\n\
+             %s"
+            total_num_failed_attempts worst
+
+  module TopThree = struct
+    let maybe_drop len = function
+      | [] ->
+          []
+      | _ :: rest as all ->
+          if len >= 3 then rest else all
+
+    (* util to maintain a list of up to 3 offending addresses - least bad to most bad *)
+    let maybe_add xs x =
+      match xs with
+      | [] ->
+          [x]
+      | x0 :: _ ->
+          let len = List.length xs in
+          if len > 3 then
+            failwith "TopThree.maybe_add: expected len <= 3"
+          else if len = 3 && x.num_failed_attempts <= x0.num_failed_attempts
+          then
+            xs
+          else
+            List.sort
+              (fun x y -> compare x.num_failed_attempts y.num_failed_attempts)
+              (x :: xs)
+            |> maybe_drop len
+  end
+
+  module Stats : sig
+    val get : unit -> stats option
+
+    (* returns the number of failures from this client since last call to [ get ] *)
+    val record_client : client -> now:Date.iso8601 -> int
+
+    (* returns number of failures from unknown clients since last call to [ get ] *)
+    val record_unknown : unit -> int
+  end = struct
+    let m = Mutex.create ()
+
+    let unknown_ctr = ref 0
+
+    let record_unknown () =
+      Mutex.execute m (fun () ->
+          let ctr = !unknown_ctr + 1 in
+          unknown_ctr := ctr ;
+          ctr)
+
+    type key = client
+
+    type value = {num_failed_attempts: int; last_failed_attempt: Date.iso8601}
+
+    type table_t = (key, value) Hashtbl.t
+
+    let table = Hashtbl.create 10
+
+    let record_client k ~now =
+      Mutex.execute m (fun () ->
```

Comment:
It looks like we are keeping a counter for all clients. Given how rarely the top3 are reported, is it not enough to query the hash table and sort the results to obtain the 3 top values?

Reply:
I would try to obtain the last 3 values when it needs to be calculated instead of all the time.